File: src/gui/backends/sdl2/application.rs
use std::{time::Duration, collections::HashMap, iter::Once, ops::DerefMut};

use sdl2::{event::Event};
use once_cell::sync::OnceCell;

use crate::gui::Color;

use super::{Window, Font};

// pub use crate::gui::backends::GenericApplication;
// pub use crate::gui::backends::GenericWindow;

// static mut TTF_CONTEXT: ManuallyDrop<sdl2::ttf::Sdl2TtfContext> = ManuallyDrop::new(sdl2::ttf::Sdl2TtfContext {});

pub struct Application<'a>
{
    pub sdl_context: sdl2::Sdl,
    pub video_subsystem: sdl2::VideoSubsystem,
    pub event_pump: sdl2::EventPump,
    pub windows: Vec<&'a mut super::Window<'a>>,
    // pub font_manager: FontManager<'a>,
    // pub fonts: HashMap<&'a str, &'a Font<'a>>,
    // pub font: &'a Font<'a>,
    // pub fonts: HashMap<String, Font<'a>>,
    // pub ttf_context: OnceCell<sdl2::ttf::Sdl2TtfContext>,
    // pub ttf_context: ManuallyDrop<sdl2::ttf::Sdl2TtfContext>,
}

impl<'a> Application<'a>
{
    pub fn new() -> Result<Application<'a>, String>
    {
        let sdl_context = sdl2::init().unwrap();
        let event_pump = sdl_context.event_pump()?;
        let video_subsystem = sdl_context.video()?;
        // Font::init()?;
        // let ttf_context = ManuallyDrop::new(sdl2::ttf::init().map_err(|e| e.to_string())?);

        Ok(Application {
            sdl_context,
            video_subsystem,
            event_pump,
            windows: vec![],
            // font,
            // font_manager: FontManager::init(),
            // fonts: HashMap::new(),
            // ttf_context: OnceCell::new(),
            // ttf_context,
        })

        // res.ttf_context.set(sdl2::ttf::init().map_err(|e| e.to_string())?).map_err(|_| "Error".to_string())?;

        // let font = res.ttf_context.get().unwrap().load_font("src/gui/assets/Roboto/Roboto-Regular.ttf", 12)?;

        // let font = res.font_manager.load("name".to_string(), "src/gui/assets/Roboto/Roboto-Regular.ttf", 12, Color::BLACK)?;

        // res.fonts.insert(font.name.as_str(), &font);
        // res.fonts.insert("font.name".to_string(), font);

        // Ok(res)
    }

    // pub fn get_font(&self) -> &'a Font
    // {
    //     self.font
    // }

    // pub fn get_font_mut(&'a mut self, font_name: &str) -> Option<&mut &'a Font<'a>>
    // {
    //     self.fonts.get_mut(font_name).map(|font| font)
    // }

    pub fn draw_window(&mut self, window: &'a mut Window<'a>)
    {
        self.windows.push(window);
    }

    pub fn main_loop(&mut self)
    {
        // self.canvas.clear();
        // self.canvas.present();
        'running: loop {
            for event in self.event_pump.poll_iter() {
                match event {
                    Event::Quit {..} /*|
                    Event::KeyDown { keycode: Some(Keycode::Escape), .. }*/ => {
                        break 'running
                    },
                    _ => {}
                }
            }
            // The rest of the game loop goes here...
            self.windows.iter_mut().for_each(|window| window.draw());

            // // put new color here
            // self.canvas.clear();
            // self.canvas.present();
            std::thread::sleep(Duration::new(0, 1_000_000_000u32 / 60));
        }
    }
}

// impl<'a> Drop for Application<'a>
// {
//     fn drop(&mut self) {
//         drop(self.windows);
//         Font::deinit();
//     }
// }

File: src/gui/backends/sdl2/application.zig
const sdl2 = @import("sdl2.zig");

const std = @import("std");

const window_mod = @import("window.zig");
const Button = @import("button.zig");
const font_mod = @import("font.zig");
const Color = @import("../../color.zig");
const Position = @import("../../position.zig");
const WindowFlags = @import("../../window_flags.zig").WindowFlags;

const allocator = std.testing.allocator;

const ApplicationError = error{
    SDL_INIT_FAILED,
    FONT_INIT_FAILED,
};

const Application = struct {
    const Self = @This();

    windows: std.ArrayList(*window_mod.Window),

    pub fn new() ApplicationError!Application {
        if (sdl2.SDL_Init(sdl2.SDL_INIT_EVERYTHING) != 0) {
            std.log.err("SDL_Init Error: {s}", .{sdl2.SDL_GetError()});
            return ApplicationError.SDL_INIT_FAILED;
        }

        if( sdl2.TTF_Init() != 0 ) {
            std.log.err("TTF_Init Error: {s}", .{sdl2.SDL_GetError()});
            return ApplicationError.FONT_INIT_FAILED;
        }

        return Application{ .windows = std.ArrayList(*window_mod.Window).init(allocator) };
    }

    pub fn add_window(self: *Self, window: *window_mod.Window) void {
        self.windows.append(window) catch undefined;
    }

    pub fn main_loop(self: *Self) void {
        var close: bool = false;

        while (!close) {
            var event: sdl2.SDL_Event = undefined;

            while (sdl2.SDL_PollEvent(&event) != 0) {
                switch (event.type) {
                    sdl2.SDL_QUIT => close = true,
                    else => {},
                }

                for (self.windows.items) |window| {
                    window.*.draw();
                }
                std.time.sleep(1_000_000_000 / 60);
            }
        }
    }

    pub fn destroy(self: Self) void {
        sdl2.SDL_Quit();
        sdl2.TTF_Quit();
        self.windows.deinit();
    }
};

test "testing application" {
    var app = try Application.new();
    defer app.destroy();

    var window = try window_mod.Window.new("hello dear", 800, 600, Color.WHITE, WindowFlags { .Shown = true }, &allocator);
    defer window.destroy();

    var font = font_mod.Font.new("src/gui/assets/Roboto/Roboto-Regular.ttf", 16, Color.BLACK) catch unreachable;
    defer font.destroy();

    var button = Button.new("button", Position.new(50, 50), Color.RED, &font, allocator);
    defer button.destroy();

    window.add_button(&button);

    app.add_window(&window);
    app.main_loop();
}
